packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -316,41 +321,43 @@ class _FullTimelineFlameChartState
         data: event,
         onSelected: (dynamic event) => widget.onSelected(event),
       );
-
       chartNodesByEvent[event] = node;
 
       rows[row].nodes.add(node);
-
-      var nextRow = row + 1;
-      for (var child in event.children) {
-        createChartNodes(child, nextRow, section);
-        if (event.hasOverlappingChildren) {
-          nextRow += child.displayDepth;
-        }
-      }
     }
 
     int currentRowIndex = _rowOffsetForTopTimeline;
     int currentSectionIndex = 0;
-    for (String bucketName in widget.data.eventBuckets.keys) {
-      final List<TimelineEvent> bucket = widget.data.eventBuckets[bucketName];
-      int sectionDepth = 0;
-      for (TimelineEvent event in bucket) {
-        resetColorOffsets();
-        sectionDepth = math.max(sectionDepth, event.displayDepth);
-        createChartNodes(event, currentRowIndex, currentSectionIndex);
+    for (String groupName in widget.data.eventGroups.keys) {
+      final FullTimelineEventGroup group = widget.data.eventGroups[groupName];
+      // Expand rows to fit nodes in [group].
+      rows.addAll(List.generate(
+        group.eventsByRow.length,
```

Comment:
this pattern of adding more rows is duplicated a couple of places. Add a helper method that implements it along with proper edge case checking. I'm not immediately clear from looking at the code why in this case we know we are always needed to add more rows to the bottom while in other cases we need checks.

Reply:
could be nice to add an assert here just to make sure the assumptions are right. That will document your assumptions and make it easier to catch if the assumptions are sometimes wrong.

Reply:
done